Terminal.Gui/Views/TabView.cs

```diff
@@ -0,0 +1,794 @@
+		}
+
+		/// <summary>
+		/// Updates <see cref="TabScrollOffset"/> to ensure that <see cref="SelectedTab"/> is visible
+		/// </summary>
+		public void EnsureSelectedTabIsVisible ()
+		{
+			if (SelectedTab == null) {
+				return;
+			}
+
+			// if current viewport does not include the selected tab
+			if (!CalculateViewport (Bounds).Any (r => Equals (SelectedTab, r.Tab))) {
+
+				// Set scroll offset so the first tab rendered is the
+				TabScrollOffset = Math.Max (0, Tabs.IndexOf (SelectedTab));
+			}
+		}
+
+		/// <summary>
+		/// Returns the number of rows occupied by rendering the tabs, this depends 
+		/// on <see cref="TabStyle.ShowHeaderOverline"/> and can be 0 (e.g. if 
+		/// <see cref="TabStyle.TabsOnBottom"/> and you ask for <paramref name="top"/>).
+		/// </summary>
+		/// <param name="top">True to measure the space required at the top of the control,
+		/// false to measure space at the bottom</param>
+		/// <returns></returns>
+		private int GetTabHeight (bool top)
+		{
+			if (top && Style.TabsOnBottom) {
+				return 0;
+			}
+
+			if (!top && !Style.TabsOnBottom) {
+				return 0;
+			}
+
+			return Style.ShowHeaderOverline ? 3 : 2;
+		}
+
+
+		/// <summary>
+		/// Returns which tabs to render at each x location
+		/// </summary>
+		/// <returns></returns>
+		private IEnumerable<TabToRender> CalculateViewport (Rect bounds)
+		{
+			int i = 1;
+
+			// Starting at the first or scrolled to tab
+			foreach (var tab in Tabs.Skip (TabScrollOffset)) {
+
+				// while there is space for the tab
+				var tabTextWidth = tab.Text.Sum (c => Rune.ColumnWidth (c));
+
+				string text = tab.Text.ToString();
+				int maxWidth = Math.Max(0,MaxTabTextWidth);
+
+				if(tabTextWidth > maxWidth){
+					text = tab.Text.ToString().Substring(0,maxWidth);
+					tabTextWidth = maxWidth;
+				}
+
+				// if there is not enough space for this tab
+				if (i + tabTextWidth >= bounds.Width) {
+					break;
+				}
+
+				// there is enough space!
+				yield return new TabToRender (i, tab,text, Equals (SelectedTab, tab), tabTextWidth);
+				i += tabTextWidth + 1;
+			}
+		}
+
+
+		/// <summary>
+		/// Adds the given <paramref name="tab"/> to <see cref="Tabs"/>
+		/// </summary>
+		/// <param name="tab"></param>
+		/// <param name="andSelect">True to make the newly added Tab the <see cref="SelectedTab"/></param>
+		public void AddTab (Tab tab, bool andSelect)
+		{
+			tabs.Add (tab);
+
+			if (SelectedTab == null || andSelect) {
+				SelectedTab = tab;
+
+				EnsureSelectedTabIsVisible ();
+
+				tab.View?.SetFocus ();
+			}
+
+			SetNeedsDisplay ();
+		}
+
+
+		/// <summary>
+		/// Removes the given <paramref name="tab"/> from <see cref="Tabs"/>.
+		/// Optionally disposes the tabs hosted <see cref="Tab.View"/>>
+		/// </summary>
+		/// <param name="tab"></param>
+		/// <param name="dispose">True to dispose of the tabs control</param>
+		public void RemoveTab (Tab tab, bool dispose)
```

Comment:
I wonder (I have no idea if my intution is correct, so please debate) if it would just be simpler for you to have the documentation say "caller is responsible for disposing any Tabs added". 

You don't normally see APIs on .NET classes that include a `dispose` option.